service/worker/scanner/executions/shard/scanner.go

```diff
@@ -0,0 +1,141 @@
+// The MIT License (MIT)
+//
+// Copyright (c) 2017-2020 Uber Technologies Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package shard
+
+import (
+	"fmt"
+
+	"github.com/pborman/uuid"
+
+	"github.com/uber/cadence/common/blobstore"
+	"github.com/uber/cadence/service/worker/scanner/executions/common"
+	"github.com/uber/cadence/service/worker/scanner/executions/invariants"
+)
+
+type (
+	scanner struct {
+		shardID          int
+		itr              common.ExecutionIterator
+		failedWriter     common.ExecutionWriter
+		corruptedWriter  common.ExecutionWriter
+		invariantManager common.InvariantManager
+	}
+)
+
+// NewScanner constructs a new scanner
+func NewScanner(
+	shardID int,
+	pr common.PersistenceRetryer,
+	persistencePageSize int,
+	blobstoreClient blobstore.Client,
+	blobstoreFlushThreshold int,
+	invariantCollections []common.InvariantCollection,
+) common.Scanner {
+	id := uuid.New()
+	return &scanner{
+		shardID:          shardID,
+		itr:              common.NewPersistenceIterator(pr, persistencePageSize, shardID),
+		failedWriter:     common.NewBlobstoreWriter(id, common.FailedExtension, blobstoreClient, blobstoreFlushThreshold),
+		corruptedWriter:  common.NewBlobstoreWriter(id, common.CorruptedExtension, blobstoreClient, blobstoreFlushThreshold),
+		invariantManager: invariants.NewInvariantManager(invariantCollections, pr),
+	}
+}
+
+// Scan scans over all executions in shard and runs invariant checks per execution.
+func (s *scanner) Scan() common.ShardScanReport {
```

Comment:
Is there any plan for adding logs & metrics for the scanner? or it will be done later in the workflow code?

Reply:
All metrics and logs will exist at the workflow level.